### DISCUSSION TITLE
Mutable strings and Racket interop

### DISCUSSION THREAD

**Discussion by samth** [racket/rhombus#292] [General]
Rhombus now expects strings to be immutable rather pervasively. This is partly because `String` only matches Racket's immutable strings, even more so because a `Map` with the default hashing compares mutable and immutable strings differently. Unfortunately that has significant conflicts with runtime features that produce strings, like regular expressions and IO. So, for example this doesn't work (using @samdphillips `rhombus/compat` library):

```
import: rhombus/compat/port
import: lib("racket/base.rkt") as racket
let stdin :: InputPort = racket.#{current-input-port}()
let keys = {"sam": 5}
let input = keys[stdin.readline()]
```

No matter what you type in, it will never match the key. Similar problems arise with regular expression splitting of strings, and with other core Racket operations that produce strings. 

I can think of a few approaches to take here (of course there could be some I haven't thought of):

1. Write Rhombus wrappers around the IO and regexp (and other) APIs, which convert from mutable to immutable strings. Using [`unsafe-string->immutable-string!`](https://docs.racket-lang.org/reference/unsafe.html#%28def._%28%28lib._racket%2Funsafe%2Fops..rkt%29._unsafe-string-~3eimmutable-string%21%29%29) can make this pretty fast, but it adds overhead to many Rhombus APIs.
2. Change the underlying primitives to produce immutable strings, and add wrappers on the Racket side. There are some compatibility concerns here about Racket programs that use the lower-level modules, plus the same drawbacks of wrappers for the Racket side, but it makes Rhombus faster.
3. Add flags throughout the various low-level APIs to specify which version to use (or duplicate the low-level APIs, which probably just pushes the flags further down). Avoids wrappers but makes the runtime code uglier and potentially slower.
4. Change the Racket APIs to just produce immutable strings. This has many virtues, and would have been the right choice at the start, but is backwards-incompatible on the Racket side. 
5. Just let the above program be wrong once it's using the relevant Rhombus APIs. I think this is a non-starter.

I don't have a strong sense of what's best here, or whether there's a better choice that I haven't thought of yet, but it seems important to start thinking about this.

**Comment by samth** [comment]
As an aside, there should probably be a clearer printing of mutable strings than `#'{"xxx"}` since that reads as an _immutable_ string.

**Comment by jackfirth** [comment]
My vote here is for option 2. It's backwards compatible while retaining the smallest legacy code and API surface possible. And I can add Resyntax rules to push people towards using immutable strings, so they don't have to remember to use the flags.

**Comment by AlexKnauth** [comment]
I'd go for option 1. Even for large strings, `unsafe-string->immutable-string!` should be a constant-time operation which will only slow down Rhombus operations relative to Racket a little. While option 2 would seem to require copying immutable strings into newly allocated mutable ones, a linear time operation in the length of the string, more time and space cost with option 2 for Racket than what option 1 would lose for Rhombus.

**Comment by mflatt** [comment]
I am also for option 1 generally, adding immutable-string variants of primitives to Racket as needed for performance.

**Reply by samth** [reply]
My biggest concern would be functions that produce data structures with strings in them, and that thus have to be traversed, such as `regexp-match`.

**Reply by mflatt** [reply]
I agree that those will require even more work. The `regexp-match` function in particular seems like something that should have a different API in Rhombus (e.g., based on names for submatches).

**Reply by jackfirth** [reply]
We could go with a hybrid between 1 and 2, where we don't add a flag to APIs that return strings directly but do add them to APIs that return strings inside structures.

**Reply by AlexKnauth** [reply]
That would be a hybrid between 1 and 3, but yeah, that makes sense.

**Reply by jackfirth** [reply]
I was thinking add a flag to the surface API, make the core API return immutable by default, and make the surface API copy to mutable if the flag isn't used.

**Comment by jackfirth** [comment]
Option 1 isn't safe if any of the mutable strings on the racket side get reused. Do we know for a fact that doesn't happen in any string APIs? I know that `(string-append)` (with no arguments) returns the same mutable empty string every time.

**Reply by AlexKnauth** [reply]
The CS implementation of `unsafe-string->immutable-string!` has a special case for the empty string so that it never sets the mutable empty string into an immutable one. I haven't looked into the BC side though

**Reply by jackfirth** [reply]
Is the empty string the _only_ time reuse ever happens?

**Reply by mflatt** [reply]
Because current APIs return mutable strings, they generally have to produce fresh strings every time, and so `unsafe-string->immutable-string!` will be ok. But certainly `unsafe-string->immutable-string!` will have to be used with some care, and `string->immutable-string` may be fast enough to make it the right choice in most cases.

**Reply by AlexKnauth** [reply]
The shared mutable empty string case is a problem on the BC side:
```racket
$ ./racket/bin/racketbc
Welcome to Racket v8.8.0.9 [bc].
> (require rackunit racket/unsafe/ops)
> (void (unsafe-string->immutable-string! (string-append)))
> (check-false (immutable? (string-append)))
--------------------
FAILURE [,bt for context]
name:       check-false
location:   string:1:0
params:     '(#t)
--------------------
```
This test passes on CS but fails on BC.

**Reply by mflatt** [reply]
That seems like bad behavior by BC, and probably `unsafe-string->immutable-string!` needs to change to cooperate with `string-append`.

**Reply by mflatt** [reply]
(Or maybe `unsafe-string->immutable-string!` should generally not mutate an empty string in BC, since other functions may has a similar optimization.)

**Reply by jackfirth** [reply]
It definitely shouldn't, since there's an advantage to ensuring there's only one immutable empty string instance in the runtime.

**Comment by sorawee** [comment]
Racket switched lists from mutable to immutable. I see no reason why we can't do the same for strings, considering that the list change presumably was larger and affected more programs.

**Reply by samth** [reply]
The Racket community was a lot smaller then, and the set of existing Racket code much easier to find and modify. I don't think we can follow that example.

**Reply by mfelleisen** [reply]
@samth Do people really program with mutable strings? I conjecture that string mutation is much rarer in the current Racket community than list mutation was in the Racket community back then. I think I'd welcome immutable strings very much.

**Reply by jackfirth** [reply]
I'd at least like to see how many programs would break before ruling it out entirely as an option.

**Reply by samdphillips** [reply]
Here are 800ish examples

https://github.com/search?q=lang%3ARacket+string-set%21&type=code

**Reply by jackfirth** [reply]
Those aren't quite breakages though, just usages of mutable strings. A breakage would be a program that relies on something like `string-append` currently returning a mutable string. Code that starts from scratch with `make-string` would be fine.

**Reply by AlexKnauth** [reply]
I had been thinking the 3 operations that would make sense to still return mutable strings would be `string-copy`, `make-string`, and `build-string`. But not all code will use one of those 3 operations before mutation. Some I see use `list->string`, `read-line`, `read`, (one used `(apply string (string->list s))` with a comment saying it was to copy it into a mutable string, why not use `string-copy` there?).

I notice even those have something in common. All of them except `string-copy` do *not* take another string as input. With that as a guideline maybe the set of operations returning mutable strings should be expanded to:
 - `make-string`
 - `string`
 - `string-copy`
 - `list->string`
 - `build-string`
 - `read`, `read-line`, `read-string`, `file->string`, `port->string`, and so on
 - `get-output-string`, `call-with-output-string`, and so on
 - `number->string`, `real->decimal-string`
 - `format`, `~a`, `~s`, `~v`, `~r`, and the rest of the formatting functions
 - `bytes->string/utf-8`, `bytes->string/locale`, `bytes->string/latin-1`
 - `symbol->string`
 - `keyword->string`
 - other various `->string` conversion functions

Perhaps the only functions that would be changed to produce immutable strings would be functional string operations that take strings a produce strings based on those, in which case maybe they should simply preserve immutability when their inputs are immutable, and produce mutable strings when one of their inputs is mutable.

**Reply by AlexKnauth** [reply]
Although, here's an example of `string-set!` being used on a string produced by `string-join`: [FPBench/src/core2cakeml.rkt:21](https://github.com/FPBench/FPBench/blob/6feb1dd753e465ac8bce7bd19f837749982ffee8/src/core2cakeml.rkt#L21-L30)

The strings inside the list passed to `string-join` are produced by `string` and `format`, so it seems good if those produce mutable strings and `string-join` only produces immutable if its inputs are immutable.

Edit: another example of `string-set!` being used on a string produced by `string-append`: [sxml/sxml/ssax/input-parse.rkt:161](https://github.com/jbclements/sxml/blob/5d1d65561b7bf5059456934c34a5c5f257de4416/sxml/ssax/input-parse.rkt#L161-L166). Again the strings inside `string-append` are produced by `make-string`, so it's good if `string-append` only produces immutable if its inputs are immutable.

**Reply by AlexKnauth** [reply]
I found an example that wouldn't be solved by making `string-append` etc. only produce immutable on immutable inputs in the Racket repo itself: [racket/racket/src/schemify/infer-name.rkt:67](https://github.com/racket/racket/blob/f6e466b4058e2343a87165b5a458bba4d3339126/racket/src/schemify/infer-name.rkt#L67-L76).
```racket
(define (source->string src)
  (define str (if (string? src) src (path->string src)))
  (define short-str
    (cond
      [((string-length str) . < . 20) (string-copy str)]
      [else (string-append "..." (substring str (- (string-length str) 19)))]))
  (for ([i (in-range (string-length short-str))])
    (when (char=? #\\ (string-ref short-str i))
      (string-set! short-str i #\/)))
  short-str)
```
Assuming `source->string` can take an immutable string as input, when that string is long (length 20 or more) it uses `string-append` and `substring` with immutable string arguments, and then it expects the result to be mutable.

Edit: I found another example that wouldn't be solved by that: [FARGish/cp1/scheme-experiments/m1.rkt:83](https://github.com/bkovitz/FARGish/blob/808e4ad95c360e88626b841254b633f5bbb941df/cp1/scheme-experiments/m1.rkt#L83). This program uses `string-append` with one argument, expecting that to convert a possibly-immutable string to a mutable string.

**Reply by samth** [reply]
While it might be easier to change multiple things all at once, having `string-append` and `substring` produce mutable strings too would not be a big issue from the Rhombus perspective. The important things to change are the places strings are produced in the runtime, as I said initially.

**Comment by mfelleisen** [comment]
“800” — that’s a small number. I expected more.

**Comment by gus-massa** [comment]
I'd like that only `string-copy` and `make-string`. I'm even not sure about `build-string`. It's extreme and it may be a backward compatibility nightmare.

Two ideas, I'm not sure about them:

1) Drop any guaranty about eq-ness of immutable strings, so the functions can return a fresh immutable string, or an old one. Even `string->immutable-string` is not guaranteed to return the same string  when the argument is immutable, it may be an interned immutable string, or just an immutable string with the same content that the compiler/optimizer has available for some weird reason. The only guaranty is that `string->immutable-string` will not do something stupid and allocate a new string when the argument is an immutable string. Somewhat like flonums. [I think this is a good idea. It is just a warning in the docs, and remove any guarantee about reuse that is there.]

2) Under the hood, the result of all internal functions are fresh mutable strings, and the compiler adds `string->immutable-string` to transform the result into an immutable string. But the compiler/optimizer is aware of the freshness of the strings, and can transform the `string->immutable-string` into `unsafe-string->immutable-string!` and remove `string-copy` and apply other local optimizations. [It needs a very smart compiler. I think it's possible, but writing the code is always more difficult than talking. Also, it needs some tricks to reuse immutable strings when possible.]

**Comment by AlexKnauth** [comment]
Inspired by the discussion on normalizing-annotations / impersonator-annotations vs flat-annotations, here's what it would currently look like in terms of impersonator contracts:
```racket
(define (string->immutable-string-proj blame)
  (λ (s neg-party)
    (cond
      [(string? s) (string->immutable-string s)]
      [else
       (raise-blame-error blame #:missing-party neg-party s
                          '(expected "string?" given: "~e")
                          s)])))

(define ->immutable-string/c
  (make-contract
   #:name '->immutable-string/c
   #:first-order string?
   #:late-neg-projection string->immutable-string-proj))
```
With the `->immutable-string/c` contract used in ways like this in a wrapper over existing racket functions that produce mutable strings, to traverse output and convert them to immutable ones:
```racket
(define/contract regexp-match/immutable
  (-> regexp?
      string?
      (or/c #f
            (cons/c ->immutable-string/c (listof (or/c ->immutable-string/c #f)))))
  regexp-match)

(define/contract regexp-match*/immutable
  (->* [regexp?
        string?]
       [#:match-select (or/c (list? . -> . (or/c any/c list?)) #f)]
       (listof (or/c ->immutable-string/c (listof (or/c #f ->immutable-string/c)))))
  regexp-match*)
```
